Picard markdups mergelibs

ewels opened the pull request:
Continuation of https://github.com/ewels/MultiQC/pull/1013

Original description by @singerj:

> **Updated MarkDuplicates.py for multiple libraries**
> The current implementation only considers the firs library of the Picard's MarkDuplicates metrics file. This pull request combines the different libraries by simply adding them up. The only exception is `PERCENT_DUPLICATION`, which is recomputed at the end.
>
> Note that also `ESTIMATED_LIBRARY_SIZE` is added up even though it is really a library feature. However, currently only the first library is used.

---

With a few tweaks I managed to get the code working on all test-data files in the repository. However, I don't want to merge into `master` just yet as this is a fairly major change in behaviour and may affect a lot of people. My main concern is for files like [this test file](https://github.com/ewels/MultiQC_TestData/blob/master/data/modules/picard/MarkDuplicates/empty_sample.txt) which may have `unknown` / `undetermined` etc and we want to ignore these? But maybe this never happens or the current default is wrong?

I think that this feature is probably well suited to an opt-in approach via a new config option, but I thought it was worth a little discussion first.

Any thoughts? @singerj / @parlundin / others?

singerj commented:
Hi @ewels ,

Sorry for the late reply. 

I totally agree with your concerns regarding the unknown/undetermined reads. However, in general I am interested in the overall statistics of the read distribution of a particular file, without caring too much which library the reads belong to. Therefore, I would include the unknown/undetermined reads. Maybe one could raise a warning, letting the user know that there is a problem? Or quit MultiQC and tell the user to specify a certain config option before relaunching? 

Also, the [test file](https://github.com/ewels/MultiQC_TestData/blob/master/data/modules/picard/MarkDuplicates/empty_sample.txt) looks a little bit strange to me. Row 7 does not contain a library name and has a "?" at the end. Also, why is there row 8 if there were no reads at all? Do I miss something?

Many thanks for taking over and continuing working on this problem!

ewels commented:
Hi @singerj,

> Also, the test file looks a little bit strange to me. Row 7 does not contain a library name and has a "?" at the end. Also, why is there row 8 if there were no reads at all? Do I miss something?

That test file comes from this issue thread: https://github.com/ewels/MultiQC/issues/402#issuecomment-281819604 - I added it as files like that used to crash MultiQC and I wanted to keep testing for it.

I've gone over your code in quite a bit of detail and ended up working on it quite a lot myself - see the last two commits. The merging behaviour is now configurable, but on by default. Libraries can be kept separate with the following MultiQC config though:

```yaml
picard_config:
    markdups_merge_multiple_libraries: False
```

To make this work I had to refactor quite a lot of the module code. I also added in functionality to recalculate the `ESTIMATED_LIBRARY_SIZE` field, which also needed recalculation in addition to the `PERCENT_DUPLICATION` field. I took this from the Picard codebase and converted to Python, it seems to work well..

Please have a go with this new version and let me know if you find any bugs!

Phil
